### PR TITLE
Add sshkey securityprovider support

### PR DESCRIPTION
**Pull request by jhg03a**
I've added support and loose documentation on getting this module to support the sshkey securityprovider plugin for mcollective in addition to ssl.  Due to time constraints and knowledge limitations, I can't provide tests at this time.  However this is pulled from my running puppet 4 environment.  Comments and tests are welcome.

There is also a small benefit for the ssl plugin in that you can now also provide the user key+cert parameters by content instead of source which would be helpful for hiera-eyaml.

### CODE REVIEW

**Comment by jyaworski** [comment]
Lints are failing. Could you run puppet-lint -f a few times?

**Comment by jhg03a** [comment]
The travis failure is expected.  The name of the ssl certificate used for the server middleware was changed from the calling user name to a generic 'server_cert.pem' because that was a more accurate reflection of what it is.  Refer to commit https://github.com/jhg03a/puppet-mcollective-1/commit/d4f79c11c57b2c1526f0ffaf5ce87adce5c7617d

**Comment by jyaworski** [comment]
This looks good. Can you squash?

**Comment by jyaworski** [comment]
Checks are failing. Can you rebase/squash/fix?

**Comment by jhg03a** [comment]
@jyaworski I messed up my repo reflog I think.  I just changed jobs officially a few hours ago, so I don't have my old environment to sort out the mess for a bit.

**Comment by jyaworski** [comment]
@jhg03a no problem. Just squash whenever you're done.

**Comment by jyaworski** [comment]
Is this good to merge? Can you squash again?

**Comment by igalic** [comment]
@jhg03a if you need some help with the [squash](https://www.git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Squashing-Commits) (and [rebase](https://www.git-scm.com/book/en/v2/Git-Branching-Rebasing)) you can take a look at https://github.com/voxpupuli/puppet-jira/pull/99#issuecomment-152528108 or come join #voxpupuli on freenode and we'll help you

**Comment by jhg03a** [comment]
I appreciate the help and I may take you up on that.  I lost my test environment, so I'm hesitant to do the merge blind.  I should hopefully get a new setup going soon so I can test this out.

**Comment by jyaworski** [comment]
Ping @jhg03a

File: manifests/init.pp [hunk 1]
```diff
@@ -94,6 +104,17 @@
   $middleware_ssl_cert_path = "${ssldir}/middleware_cert.pem"
   $middleware_ssl_ca_path   = "${ssldir}/middleware_ca.pem"
 
+  if $securityprovider == 'sshkey' {
+    package{'sshkeyauth':
```
Comment (hunner) [inline manifests/init.pp]: maybe spaces around `{`

File: manifests/init.pp [hunk 2]
```diff
@@ -94,6 +104,17 @@
...
+      ensure   =>  $sshkeyauth_gem_version,
+      provider =>  'puppet_gem',
```
Comment (hunner) [inline manifests/init.pp]: These have an extra space after the `=>`

File: manifests/init.pp [hunk 3]
```diff
@@ -94,6 +104,17 @@
...
+    }
+  }
+
+  if $sshkey_server_learn_public_keys {
+    $sshkey_server_publickey_dir_real = pick($sshkey_server_publickey_dir,"${confdir}/sshkey_pubkeys")
```
Comment (hunner) [inline manifests/init.pp]: Instead of `$*_real` we usually use `$_*` for intermediary variables.

File: manifests/init.pp (hunk 2 above)
Comment (bmjen) [inline manifests/init.pp]: This needs to be conditionalized. I don't think this provider will work on Puppet <4.0. Provider for Puppet 3.x should be 'gem'.

File: manifests/user.pp [hunk 4]
```diff
@@ -5,41 +5,65 @@
   $group       = $name,
   $homedir     = "/home/${name}",
   $certificate = undef,
+  $certificate_content  = undef,
   $private_key = undef,
+  $private_key_content  = undef,
+  $public_key = undef,
+  $public_key_content  = undef,
```
Comment (hunner) [inline manifests/user.pp]: Align arrows

File: manifests/user.pp [hunk 5]
```diff
@@ -5,41 +5,65 @@
...
+# Validate that both forms of data weren't given
+  if $certificate and $certificate_content {
+    fail("Both a source and content cannot be defined for ${username} certificate!")
+  }
+  if $private_key and $private_key_content {
+    fail("Both a source and content cannot be defined for ${username} private key!")
+  }
+
+  # Variable interpolation in class parameters can be goofy (PUP-1080)
+  $homedir_real              = pick($homedir,"/home/${username}")
```
Comment (hunner) [inline manifests/user.pp]: `$_homedir` also

**Comment by igalic** [comment]
thanks @hunner and @bmjen for reviewing this
if @jhg03a doesn't step up, can anyone else maybe follow up on these comments?

**Pull request closed without merging**